fp-core/src/foldable.rs

```diff
@@ -1,17 +1,49 @@
 use crate::hkt::HKT;
 use crate::monoid::Monoid;
 
-pub trait Foldable<A, B>: HKT<A, B> {
-    fn reduce<F>(b: B, ba: F) -> <Self as HKT<A, B>>::Target
+// Cheating: all HKT instances exist for any B,
+// so HKT<B> here isn't about Self<B> having any meaning,
+// it's about folding on some Self<A> -- the HKT lets us
+// have an A to speak of.
+pub trait Foldable<B>: HKT<B> + Sized {
+    fn reduce<F>(self, b: B, ba: F) -> B
     where
-        F: FnOnce(B, A) -> (B, B);
+        F: Fn(B, &<Self as HKT<B>>::Current) -> B;
```

Comment:
Following up on the chat, why do you prefer `Fn` over `FnOnce`?

Reply:
Also good catch on `-> B`

Reply:
In regards to `F: Fn(B, &<Self as HKT<B>>::Current) -> B;`, shouldn't it return (accumlation, next_value)? so I was thinking returning (B, B) is right.

Reply:
Wrt to fn: https://doc.rust-lang.org/std/ops/trait.Fn.html
I particularly like the part that says they an Fn can be called concurrently. That could help with the `future` stuff.

Reply:
Also, you don't return a `(B, B)` since the folding function has no notion of a "next value"... It doesn't know if the container is empty or it if it's at the end of the container.

You can look at `Iterator`s fold for a rust example.

Reply:
Repeating myself here ;) Let's revisit the idea of using `Self as HKT<B>>::Current`